### ISSUE TITLE
[Due for payment 2025-03-10] [$250] Security - Unable to remove closed account from copilot

### ISSUE THREAD

**Issue by lanitochka17** [Expensify/App#53785]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 9.0.73-0
**Reproducible in staging?:** Y
**Reproducible in production?:** Y
**If this was caught on HybridApp, is this reproducible on New Expensify Standalone?:** N/A
**If this was caught during regression testing, add the test name, ID and link from TestRail:** N/A
**Email or phone of affected tester (no customers):** applausetester+bp2510@applause.expensifail.com
**Issue reported by:** Applause - Internal Team

## Action Performed:
Precondition: copilot account B is added to account A. Account B was closed.
1. Account A: go to Security
2. Click on three dots near copilot account B
3. Click on Remove copilot > Remove copilot

## Expected Result:
Account B is remover from the list

## Actual Result:
Account B is no removed from the list

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Standalone
- [ ] Android: HybridApp
- [ ] Android: mWeb Chrome
- [ ] iOS: Standalone
- [ ] iOS: HybridApp
- [ ] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

<details>
  <summary>Add any screenshot/video evidence</summary>

  
</details> 

https://github.com/user-attachments/assets/f7946840-dd12-4b5d-abb1-12e0d27c90fb

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~021866325124686695861</li>
        <li>Upwork Job ID: 1866325124686695861</li>
        <li>Last Price Increase: 2024-12-17</li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @VictoriaExpensify</details>

**Comment by Shahidullah-Muffakir** [comment]
I couldn't reproduce it.

https://github.com/user-attachments/assets/56203355-aaab-48f8-8d26-fecf3a635aa6

**Comment by lanitochka17** [comment]
@Shahidullah-Muffakir please check again preconditions have been added:
Precondition: copilot account B is added to account A. Account B was closed.

**Comment by Shahidullah-Muffakir** [comment]
> @Shahidullah-Muffakir please check again preconditions have been added: Copilot account B is added to account A. Account B was closed.

Thank you, Yes I could reproduce it now.
It seems like a BE issue:

<img width="708" alt="Screenshot 2024-12-10 at 3 46 26 AM" src="https://github.com/user-attachments/assets/016ed855-71e6-462d-a53a-6cf671fe94bf">

**Comment by VictoriaExpensify** [comment]
Yeah this is an issue we should fix.

**Comment by mollfpr** [comment]
I agree this is a BE issue. The ND should send a pusher message to update the `account.delegates` to their delegators when closing an account. But this doesn't solve this issue, until the user re-login.

The other thing we can do in the ND is, call an API that contains fresh `account` data when the user opens the security page. It does solve the issue and the user does not need to re-login to get the correct delegate data.

@VictoriaExpensify We might need an internal engineer to choose the right path.

**Comment by mollfpr** [comment]
@daledah That's a good suggestion to fix the show error message, but I think the issue persists. The user still can't remove the copilot until re-login.

The pusher fix only solved the issue when the copilot closed their account.

**Comment by mollfpr** [comment]
@VictoriaExpensify Could you add the internal label? Thanks!

**Comment by mountiny** [comment]
cc @dangrous

**Comment by dangrous** [comment]
Hi! @mollfpr bumping the question above when you have a moment - let me know what we should be updating in the BE and I can take a stab at it

**Comment by dangrous** [comment]
> Hi all! hopping in in the new year. Can you confirm what we need to do on the backend? Reading through, it sounded like we needed to send an error from the BE, but I think we're already doing that, right? Let me know what I'm missing, though, and happy to tackle the BE update.

@mollfpr Bumping this - let me know what action I can take to help move this along. Thanks!

**Comment by mollfpr** [comment]
Sorry @dangrous I lost track of the issue 🙏

Here's what I think we can do in BE.

>I agree this is a BE issue. The ND should send a pusher message to update the account.delegates to their delegators when closing an account. But this doesn't solve this issue, until the user re-login.
>
>The other thing we can do in the ND is, call an API that contains fresh account data when the user opens the security page. It does solve the issue and the user does not need to re-login to get the correct delegate data.

**Comment by dangrous** [comment]
Ah okay so something sort of like the `openPolicyCompanyCardsPage` we already have? Do we do anything like that for profile info yet? Not a deal-breaker, but just wondering. I guess that's probably the easiest way, though we'd need to add a loader for the page.

**Comment by mollfpr** [comment]
Yup, similar to that or OpenPaymentsPage!

> Do we do anything like that for profile info yet?

We only have `OpenInitialSettingsPage` but it doesn't return any data related to account delegates.

**Comment by dangrous** [comment]
Okay so I've made a [draft command](https://github.com/Expensify/Web-Expensify/pull/45209) that the App will be able to call - OpenSecuritySettingsPage - that will return the following onyx update:

```
            [
                'onyxMethod' => Onyx::METHOD_MERGE,
                'key' => OnyxKeys::ACCOUNT,
                'value' => [
                    'delegatedAccess' => [up to date delegated access data],
                ],
            ],
```

Does that feel right? If so, are you up for making an App PR that calls that on opening the security page, and then I can test locally to see if that works out?

**Comment by dangrous** [comment]
bump on this ^^ when you have a moment @mollfpr - thanks!

**Comment by VictoriaExpensify** [comment]
Not overdue, waiting for @mollfpr to chime in

**Comment by mollfpr** [comment]
@dangrous Here's the draft PR https://github.com/Expensify/App/pull/55709

**Comment by dangrous** [comment]
Perfect, thanks! Looks like BE is working well. Going to clean up that PR and get that merged, then we can move ahead with yours!

**Comment by mollfpr** [comment]
@dangrous Thank you! I'll try on my PR!

**Comment by dangrous** [comment]
How's this one looking?

**Comment by mollfpr** [comment]
@dangrous I found a case that make the issue still reproduced.

While the account A is open the security page, it will show the account B as the copilot. Open the account B on other device and close the account. In this step the account B is still appear on account A and will fail if try to remove the account B from copilot.

I'm thinking to calling the API again when it happen, so it will update the copilot list. What do you think?

**Comment by mollfpr** [comment]
@dangrous Actually, after the remove delegate is failing.

>I could potentially update the backend to just "succeed" if the account is closed?

I'm onboard with this idea!

**Comment by dangrous** [comment]
Cool - backend PR with that change is in review! If a user tries to remove a delegate with a closed account, it won't error, and it'll queue an onyx update sending the updated delegatedaccess

**Comment by dangrous** [comment]
Okay, that change is live on prod - let me know if we need to adjust anything else, @mollfpr

**Comment by mollfpr** [comment]
@dangrous Thank you! The test looks great and I'll finish the PR now.

**Comment by mollfpr** [comment]
@dangrous On second thoughts, I think we don't need ND PR. There's no error in removing the close account as described in the issue.

https://github.com/user-attachments/assets/a3dee38e-5bbc-4a17-9021-29b13a1dd144

**Comment by dangrous** [comment]
I think we should still update the list on opening the security settings page, though - that'll provide the best experience, since it'll lessen the amount of times people even have to remove a delegate.

**Comment by mollfpr** [comment]
@dangrous I open a new PR https://github.com/Expensify/App/pull/56900

**Comment by VictoriaExpensify** [comment]
Payment summary: 

Reviewer: @Ollyws owed $250 via NewDot
